src/models/base_model.py

```diff
@@ -0,0 +1,226 @@
+        """
+        output = self._feature_outputs[device]
+        # clear after reading
+        self._feature_outputs[device] = OrderedDict()  
+        return output
+
+
+class BaseModel(nn.Module):
+    """Base model for all Torchok Models - Backbone, Neck, Pooling and Head.
+
+    This class support methods for add hooks.
+    Also it's have channels and reduction for input and output tensors, for each forward and hooks case. 
+    Channels - tensor channels.
+    Reduction -  downscale value for get tensor shape.
+
+    Args:
+        input_forward_channels: Input channels list for forward method.
+        input_forward_reductions: Optional value of input reduction list for forward. Need to reconstruct initial input
+            tensor shape.
+        input_hooks_channels: Optional value of input channels list for hooks.
+        input_hooks_reductions: Optional value of input hooks reduction list. Need to reconstruct initial input tensor 
+            shape.
+    """
+    def __init__(self, input_forward_channels: List[int], input_forward_reductions: List[int] = None, \
+                 input_hooks_channels: List[int] = None, input_hooks_reductions: List[int] = None):
+        super().__init__()
+        self._feature_info = None
+
+        self._input_forward_channels = input_forward_channels
+        self._input_forward_reductions = input_forward_reductions
+        self._input_hooks_channels = input_hooks_channels
+        self._input_hooks_reductions = input_hooks_reductions
+
+        self._output_forward_channels = None
+        self._output_forward_reductions = None
+        self._output_hooks_channels = None
+        self._output_hooks_reductions = None
+
+    def create_hooks(self):
+        """Generate feature hooks."""
+        # Check if all self._feature_info is FeatureInfo types.
+        self.check_feature_info_type(self)
+        self._stage_names = [feature.module_name for feature in self._feature_info]
+        self._output_hooks_channels = [feature.channel_number for feature in self._feature_info]
+        self._output_hooks_reductions = [feature.reduction for feature in self._feature_info]
+        hooks = [Hook(module_name=name, hook_type=HookType.FORWARD) for name in self._stage_names]
+        self._feature_hooks = FeatureHooks(hooks, self.named_modules())
+
+    def check_feature_info_type(self):
+        """Check if all self._feature_info is FeatureInfo class.
+            
+        Raises:
+            ValueError: If any value in feature_info is not FeatureInfo class.
+        """
+        for feature in self._feature_info:
+            if type(feature) != FeatureInfo:
+                raise ValueError('All feature_info must be FeatureInfo class.')
+
+    def forward(self, x: torch.Tensor):
+        # Return features for classification.
+        raise NotImplementedError
+
+    def forward_backbone_features(self, x: torch.Tensor) -> Tuple[torch.Tensor, List[torch.Tensor]]:
```

Comment:
Since the model is just universal `BaseModel` there shouldn't be word `backbone`. So I suggest you to join methods `forward_backbone_features` and `forward_stage_features` into a single one and call it `forward_stage_features`